analyzetestgroups

```diff
@@ -0,0 +1,252 @@
+            with open(problempath / 'problem.yaml', encoding = 'utf-8') as file:
+                problemtype = yaml.safe_load(file).get('type')
+                if problemtype != 'scoring':
+                    exit(f'FATAL: {problemname} is is not a scoring problem. Aborting...')
+            print (' '*80, end = '\r')
+            print(f'\033[01mScoring problem: {problemname}\033[0m')
+        elif p == 'START_SUBMISSION':
+            s = Submission(d['type'], d['filename'])
+        elif p == 'START_TESTGROUP':
+            tc_times: List[float] = []
+        elif p == 'AC_TC_RESULT':
+            print (problemname, s, end ='\r')
+            tc_times.append(float(d['time']))
+            tc_id = d['case']
+        elif p == 'TESTGROUP_GRADE':
+            grade = d['grade']
+            if d['type'] == 'sample': 
+                assert s is not None and len(s.verdicts) == 0
+            else:  
+                assert s is not None and len(s.verdicts) == int(d['number'])
+                assert grade != 'AC' or len(tc_times) # secret test group must have at least one tc for AC
+            s.verdicts.append(Verdict(grade, max(tc_times) if len(tc_times) else None))
+        elif p == 'END_SUBMISSION':
+            assert s is not None
+            s.points = int(d["points"] or "0")
+            s.maxtime = float(d["maxtime"])
+            submissions.append(s)
+        elif p == 'TIMELIMIT':
+            timelimits = int(d['limit']), int(d['safety'])
+        statusline = f'Submission {s}, test case {tc_id}'
+        print (' '*80, end = '\r')
+        print (statusline[:80], end = '\r')
+
+    assert timelimits is not None 
+    return timelimits, submissions
+
+def pretty_yn(value: bool) -> str:
+    # green y or red n
+     return '\033[32my\033[0m' if value else '\033[91mn\033[0m'
+
+def print_table(submissions):
+    alignto = max (len(str(sub)) for sub in submissions)
+    if alignto < len('Submission'):
+        alignto = len('Submission')
+    print("\033[01m", end = '')
+    print(f"{'Submission':{alignto}} Sample  ", end = ' ')
+    print(' '.join(f'Group {i} ' for i in range(1, num_secret_groups + 1)), end = ' ')
+    print('Pts Time  Expected\033[0m')
+    for s in submissions:
+        print (f"{s.filename:{alignto}}", end = ' ')
+        for verdict in s.verdicts:
+            print (f'{verdict:17}', end = ' ')
+        print(f"{s.points:3}", end = ' ')
+        print(f"{s.maxtime:4.2f}s", end = ' ')
+
+        # get the expected grades from the file (if exist), default to AC for AC submissions
+        expected_grades = None
+        path = problempath / 'submissions' / s.fullpath
+        with open(path, encoding = 'utf-8') as sourcefile:
+            for line in sourcefile:
+                m = expected_score_pattern.search(line)
+                if m:
+                    expected_grades = m.group("grades").split()
+                    assert len(expected_grades) == num_secret_groups
```

Comment:
Make this a properly formatted error, not an assert

Same goes for most other asserts. Assertion failures are generally only intelligible to the author of the program, I've found.

Reply:
Done in 2423e8d